File: clang/tools/metal-as/metal-as.py
#!/usr/bin/env python3
"""metal-as: Assemble LLVM bitcode into .air files with MTLB header.

This tool takes LLVM bitcode (.bc) as input and produces .air files
by adding the MTLB header that Apple's Metal compiler uses.
"""

import argparse
import struct
import sys
from pathlib import Path


def create_mtlb_header(bitcode_size: int, triple: str) -> bytes:
    """Create MTLB header for .air file.
    
    MTLB header structure (284 bytes):
    - Magic: "MTLB" (4 bytes)
    - Version/flags (4 bytes)
    - Various fields (276 bytes)
    """
    header = bytearray(284)
    
    # Magic number
    header[0:4] = b'MTLB'
    
    # Version/flags (example: 0x00028001)
    struct.pack_into('<I', header, 4, 0x00028001)
    
    # File size field (offset 16-23)
    total_size = len(header) + bitcode_size + len(triple) + 1
    struct.pack_into('<Q', header, 16, total_size)
    
    # Other fields (based on reverse engineering)
    struct.pack_into('<Q', header, 8, 4)  # Some count/flag
    struct.pack_into('<Q', header, 24, 0x82)  # Some offset/size
    struct.pack_into('<Q', header, 32, 0xf8)
    struct.pack_into('<Q', header, 40, 8)
    struct.pack_into('<Q', header, 48, 0x100)
    
    return bytes(header)


def assemble_air(bitcode_path: Path, output_path: Path, triple: str) -> None:
    """Assemble .bc file into .air file with MTLB header."""
    
    # Read LLVM bitcode
    bitcode = bitcode_path.read_bytes()
    
    # Verify it's LLVM bitcode
    if bitcode[:4] != b'BC\xc0\xde':
        print(f"Warning: {bitcode_path} does not appear to be LLVM bitcode", file=sys.stderr)
    
    # Create MTLB header
    header = create_mtlb_header(len(bitcode), triple)
    
    # Write .air file
    with open(output_path, 'wb') as f:
        f.write(header)
        f.write(bitcode)
        # Append triple string (null-terminated)
        f.write(triple.encode('utf-8'))
        f.write(b'\x00')
    
    print(f"Created {output_path} ({len(header) + len(bitcode) + len(triple) + 1} bytes)")


def main() -> None:
    parser = argparse.ArgumentParser(description='Assemble LLVM bitcode into .air files')
    parser.add_argument('input', type=Path, help='Input LLVM bitcode file (.bc)')
    parser.add_argument('-o', '--output', type=Path, required=True, help='Output .air file')
    parser.add_argument('--triple', default='air64-apple-macosx10.15.0',
                       help='Target triple (default: air64-apple-macosx10.15.0)')
    
    args = parser.parse_args()
    
    if not args.input.exists():
        print(f"Error: Input file {args.input} not found", file=sys.stderr)
        sys.exit(1)
    
    assemble_air(args.input, args.output, args.triple)


if __name__ == '__main__':
    main()

File: llvm/tools/metal-as/metal-as.cpp
//===- metal-as.cpp - Assemble LLVM bitcode into .air files ---------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
//
// This tool takes LLVM bitcode (.bc) as input and produces .air files
// by adding the MTLB header that Apple's Metal compiler uses.
//
//===----------------------------------------------------------------------===//

#include "llvm/ADT/StringRef.h"
#include "llvm/Bitcode/BitcodeReader.h"
#include "llvm/Bitcode/BitcodeWriter.h"
#include "llvm/IR/LLVMContext.h"
#include "llvm/IR/Module.h"
#include "llvm/Support/CommandLine.h"
#include "llvm/Support/FileSystem.h"
#include "llvm/Support/MemoryBuffer.h"
#include "llvm/Support/raw_ostream.h"
#include "llvm/Support/ToolOutputFile.h"
#include "llvm/Support/WithColor.h"

#include <cstring>

using namespace llvm;

static cl::opt<std::string> InputFilename(cl::Positional, cl::desc("<input bitcode>"),
                                          cl::Required);

static cl::opt<std::string> OutputFilename("o", cl::desc("Output filename"),
                                           cl::value_desc("filename"),
                                           cl::Required);

static cl::opt<std::string> TargetTriple("triple", cl::desc("Target triple"),
                                         cl::init("air64-apple-macosx10.15.0"));

static void createMTLBHeader(raw_ostream &OS, size_t BitcodeSize,
                             StringRef Triple) {
  // MTLB header structure (284 bytes)
  // Based on reverse engineering of Apple's .air files
  
  char Header[284] = {0};
  
  // Magic number
  std::memcpy(Header, "MTLB", 4);
  
  // Version/flags (little-endian)
  uint32_t Version = 0x00028001;
  std::memcpy(Header + 4, &Version, 4);
  
  // Various fields (based on reverse engineering)
  uint64_t Field1 = 4;
  std::memcpy(Header + 8, &Field1, 8);
  
  uint64_t TotalSize = sizeof(Header) + BitcodeSize + Triple.size() + 1;
  std::memcpy(Header + 16, &TotalSize, 8);
  
  uint64_t Field2 = 0x82;
  std::memcpy(Header + 24, &Field2, 8);
  
  uint64_t Field3 = 0xf8;
  std::memcpy(Header + 32, &Field3, 8);
  
  uint64_t Field4 = 8;
  std::memcpy(Header + 40, &Field4, 8);
  
  uint64_t Field5 = 0x100;
  std::memcpy(Header + 48, &Field5, 8);
  
  OS.write(Header, sizeof(Header));
}

int main(int argc, char **argv) {
  cl::ParseCommandLineOptions(argc, argv, "Metal assembler\n");
  
  // Read input bitcode
  ErrorOr<std::unique_ptr<MemoryBuffer>> MBOrErr =
      MemoryBuffer::getFileOrSTDIN(InputFilename);
  if (!MBOrErr) {
    WithColor::error() << "Could not open input file: " << MBOrErr.getError().message() << "\n";
    return 1;
  }
  
  std::unique_ptr<MemoryBuffer> MB = std::move(MBOrErr.get());
  
  // Verify it's LLVM bitcode
  if (MB->getBufferSize() < 4 ||
      std::memcmp(MB->getBufferStart(), "BC\xc0\xde", 4) != 0) {
    WithColor::warning() << "Input does not appear to be LLVM bitcode\n";
  }
  
  // Create output file
  std::error_code EC;
  std::unique_ptr<ToolOutputFile> Out =
      std::make_unique<ToolOutputFile>(OutputFilename, EC, sys::fs::OF_None);
  if (EC) {
    WithColor::error() << EC.message() << '\n';
    return 1;
  }
  
  // Write MTLB header
  createMTLBHeader(Out->os(), MB->getBufferSize(), TargetTriple);
  
  // Write bitcode
  Out->os().write(MB->getBufferStart(), MB->getBufferSize());
  
  // Write triple (null-terminated)
  Out->os().write(TargetTriple.data(), TargetTriple.size());
  Out->os().write('\0');
  
  Out->keep();
  
  outs() << "Created " << OutputFilename << " ("
         << (284 + MB->getBufferSize() + TargetTriple.size() + 1)
         << " bytes)\n";
  
  return 0;
}
